### PR TITLE
feat(integrations): VSTS Global Plugin Authorization

### PR DESCRIPTION
**Pull request by lauryndbrown**

### CODE REVIEW

File: src/sentry/integrations/vsts/urls.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from __future__ import absolute_import, print_function
+
+from django.conf.urls import patterns  # , url
+
+urlpatterns = patterns(
+    '',
+    # url(r'^action/$', SlackActionEndpoint.as_view()),
```
Comment (lauryndbrown) [inline src/sentry/integrations/vsts/urls.py]: @macqueen didn't get to the part where custom URLs were relevant.

File: src/sentry/integrations/vsts/integration.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        }]
+
+    def build_integration(self, state):
+        data = state['identity']['data']
+
+        scopes = sorted(self.identity_oauth_scopes)
+
+        return {
+            'name': 'VSTS',
+            'external_id': 'vsts',
```
Comment (lauryndbrown) [inline src/sentry/integrations/vsts/integration.py]: @macqueen `'name'` and `'external_id'` are placeholders to be team/project names that are personalized for the user. there is a concept of a default collection and you probably need to say something like this: `curl -u {username}[:{personalaccesstoken}] https://{account}.VisualStudio.com/DefaultCollection/_apis/projects?api-version=2.0` this https://docs.microsoft.com/en-us/rest/api/vsts/?view=vsts and https://docs.microsoft.com/en-us/rest/api/vsts/core/projects/list should be helpful.

File: src/sentry/identity/vsts/provider.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    </form>
+    '''
+
+    def dispatch(self, request, pipeline):
+        if 'instance' in request.POST:
+            pipeline.bind_state('instance', request.POST.get('instance'))
+            if 'default_project' in request.POST:
+                pipeline.bind_state('default_project', request.POST.get('default_project'))
+                return pipeline.next_step()
+        return HttpResponse(self.TEMPLATE)
```
Comment (lauryndbrown) [inline src/sentry/identity/vsts/provider.py]: Would this be the correct place to request the projects and give the user a dropdown? I was going to try that and then thought for now just typing in the information is fine. It can be followed up with another pr.

**Comment by lauryndbrown** [comment]
vsts_oath

**Comment by lauryndbrown** [comment]
oauth_vsts_larger

File: src/sentry/integrations/vsts/integration.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                'domain_name': instance,
+                # icon doesn't appear to be possible
+            },
+            'user_identity': {
+                'type': 'vsts',
+                'external_id': instance,
+                'scopes': [],
+                'data': {},
+            }
+        }
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/vsts/integration.py]: We probably actually want to store the access token here as a user identity, and we'll have have to find a way to pass along the default identity here for the organization integration, since this doesn't have a global access token.

File: src/sentry/integrations/vsts/integration.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        default_project_name = state['identity']['default_project']
+
+        scopes = sorted(self.identity_oauth_scopes)
+        projects = self.get_projects(instance, access_token)['value']
+        default_project = self.get_default_project(default_project_name, projects)
+        return {
+            'name': default_project['name'],
+            'external_id': default_project['id'],
+            'metadata': {
+                'access_token': access_token,
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/vsts/integration.py]: The integration won't have an access token, it will have a default identity that it can use, since the access token is associated to a user, *not* the integration itself.

File: src/sentry/identity/vsts/provider.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        body = safe_urlread(req)
+        if req.headers['Content-Type'].startswith('application/x-www-form-urlencoded'):
+            return dict(parse_qsl(body))
+        return json.loads(body)
+
+
+class ConfigView(PipelineView):
+    TEMPLATE = '''
+    <form method="POST">
+        Instance: <input type="text" name="instance" placeholder="example.visualstudio.com" required=True />
```
[the remark this answers is not in the document]
Comment (lauryndbrown) [reply src/sentry/identity/vsts/provider.py]: Unfortunately yes. :/ I had Elizabeth ask the VSTS team some questions just in case. But so far the answer is no.

File: src/sentry/integrations/vsts/integration.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    source_url='https://github.com/getsentry/sentry/tree/master/src/sentry/integrations/vsts',
+    aspects={},
+)
+
+
+class VSTSIntegration(Integration):
+    key = 'vsts'
+    name = 'VSTS'
+    metadata = metadata
+    domain = '.visualstudio.com'
```
[the remark this answers is not in the document]
Comment (lauryndbrown) [reply src/sentry/integrations/vsts/integration.py]: @EvanPurkhiser the `server:port` is for on-prem only. Since it isn't a priority, I will take that out of the description. For the off the shelf version you must have that domain.

File: src/sentry/identity/vsts/provider.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lauryndbrown) [reply src/sentry/identity/vsts/provider.py]: this is the response as per the documentation that you can find here:https://docs.microsoft.com/en-us/vsts/integrate/get-started/authentication/oauth?view=vsts

```
{
    "access_token": { access token for the user },
    "token_type": { type of token },
    "expires_in": { time in seconds that the token remains valid },
    "refresh_token": { refresh token to use to acquire a new access token }
}
```

File: src/sentry/identity/vsts/provider.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                authorize_url=self.oauth_authorize_url,
+                client_id=self.get_oauth_client_id(),
+                scope=' '.join(self.get_oauth_scopes()),
+            ),
+            VSTSOAuth2CallbackView(
+                access_token_url=self.oauth_access_token_url,
+                client_id=self.get_oauth_client_id(),
+                client_secret=self.get_oauth_client_secret(),
+            ),
+            ConfigView(),
```
[the remark this answers is not in the document]
Comment (lauryndbrown) [reply src/sentry/identity/vsts/provider.py]: Somewhat. In order to make any other requests than oauth, I need the account name (or server instance for on-prem). Unfortunately because the oauth doesn't give me anything else, I am having to ask the user for their account name (at the very least).

File: src/sentry/integrations/vsts/integration.py (hunk 5 above)
Comment (lauryndbrown) [reply src/sentry/integrations/vsts/integration.py]: Good catch. Thanks!

File: src/sentry/identity/vsts/provider.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+from sentry.utils.pipeline import PipelineView
+
+options.register('vsts.client-id', flags=FLAG_PRIORITIZE_DISK)
+options.register('vsts.client-secret', flags=FLAG_PRIORITIZE_DISK)
+options.register('vsts.verification-token', flags=FLAG_PRIORITIZE_DISK)
+
+
+class VSTSIdentityProvider(OAuth2Provider):
+    key = 'vsts'
+    name = 'VSTS'
```
[the remark this answers is not in the document]
Comment (lauryndbrown) [reply src/sentry/identity/vsts/provider.py]: I assume you mean for the name. I'd say the key is probably ok being vsts. But I agree with the name

File: src/sentry/integrations/vsts/integration.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                'help': (
+                    'Enter the Visual Studio Team Services project name that you wish '
+                    'to use as a default for new work items'
+                ),
+            },
+        ]
+
+    def get_projects(self, instance, access_token):
+        session = http.build_session()
+        url = 'https://%s/DefaultCollection/_apis/projects' % instance
```
[the remark this answers is not in the document]
Comment (EvanPurkhiser) [reply src/sentry/integrations/vsts/integration.py]: Yeah, I just always use unicode strings with `format`. It's fine.

[the remark this answers is not in the document]
Comment (mattrobenolt) [reply src/sentry/integrations/vsts/integration.py]: Yeah, if you use `unicode.format` you're generally fine, but `str.format` causes lot of issues and people typically don't explicitly do that. It has been removed in python 3, assuming because 1, it doesn't make a lot of sense, and 2, causes problems.

Using `unicode.format` requires prefixing your string explicitly with `u` to make it behave that way.

**Comment by lauryndbrown** [comment]
vsts-start
vsts-microsoft-auth
vsts-account
vsts-project
vsts-project-selection
vsts-success

File: src/sentry/conf/server.py [hunk 11]
```diff
@@ -1237,6 +1237,11 @@ def create_partitioned_queues(name):
 SENTRY_DEFAULT_INTEGRATIONS = (
     'sentry.integrations.slack.SlackIntegration',
     'sentry.integrations.jira.JiraIntegration',
+    'sentry.integrations.vsts.VSTSIntegration',
+)
+
+SENTRY_INTERNAL_INTEGRATIONS = (
+    'vsts',
```
Comment (EvanPurkhiser) [inline src/sentry/conf/server.py]: Nice, can you add `github` and `jira` too?

File: src/sentry/identity/vsts/provider.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+from sentry import options
+
+from sentry.web.helpers import render_to_response
+from sentry.options.manager import FLAG_PRIORITIZE_DISK
+from sentry.identity.oauth2 import OAuth2Provider, OAuth2LoginView, OAuth2CallbackView
+from sentry.pipeline import PipelineView
+
+options.register('vsts.client-id', flags=FLAG_PRIORITIZE_DISK)
+options.register('vsts.client-secret', flags=FLAG_PRIORITIZE_DISK)
+options.register('vsts.verification-token', flags=FLAG_PRIORITIZE_DISK)
```
Comment (EvanPurkhiser) [inline src/sentry/identity/vsts/provider.py]: Lets moves these into `sentry/options/defaults.py`

File: src/sentry/identity/vsts/provider.py [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+        'vso.release_manage',
+        'vso.security_manage',
+        'vso.serviceendpoint_manage',
+        'vso.symbols_manage',
+        'vso.taskgroups_manage',
+        'vso.test_write',
+        'vso.wiki_write',
+        'vso.work_full',
+        'vso.workitemsearch',
+    )
```
Comment (EvanPurkhiser) [inline src/sentry/identity/vsts/provider.py]: Can you cut these down to just what we need for identity verification? Maybe no scopes initially? The identity pipeline usually just has scopes to obtain the users identity.

File: src/sentry/integrations/vsts/integration.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
+from __future__ import absolute_import
+from sentry import http
+
+from django import forms
+from sentry.web.helpers import render_to_response
+from sentry.integrations import Integration, IntegrationMetadata
+from sentry.pipeline import NestedPipelineView, PipelineView
+from sentry.identity.pipeline import IdentityProviderPipeline
+from sentry.utils.http import absolute_uri
+DESCRIPTION = """
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/vsts/integration.py]: Doesn't the formatter complain about missing spaces after the imports?

File: src/sentry/integrations/vsts/integration.py [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+        return {
+            'name': project['name'],
+            'external_id': project['id'],
+            'metadata': {
+                'domain_name': instance,
+                'scopes': scopes,
+                # icon doesn't appear to be possible
+            },
+            'user_identity': {
+                'access_token': access_token,
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/vsts/integration.py]: Access token actually goes in the `data` object, sorry should have made that more clear. Check out the `get_oauth_data` method used in the slack identity provider to get the oauth data key (Would be nice if we could reuse this, but would need some refactoring).

https://github.com/getsentry/sentry/blob/c62a7b251ee9fc840a81621374ba34bf46a7ceb5/src/sentry/identity/oauth2.py#L56-L66

Comment (EvanPurkhiser) [reply src/sentry/integrations/vsts/integration.py]: This is definitely a bit different from the Slack integration, since fundamentally the access token that comes back from slack is for an entire workspace, where as this token is for a users account.

**Review by evanpurkhiser — COMMENTED** [review]
Looking really good, just a few minor changes. Ping me in person if you need any clarification :)

File: src/sentry/identity/vsts/provider.py (hunk 13 above)
Comment (lauryndbrown) [reply src/sentry/identity/vsts/provider.py]: @EvanPurkhiser The scopes must match the configuration setup in the application. Because this is the step users give the permissions to the application. And once set on VSTS side, there doesn't appear to be a way to alter them after the fact. So, rather than recreating the application each time I needed a scope change, I decided to just grab any scope we could reasonably want, and we could pair down after the functionality was decided upon. 

I'd say if you'd rather we went with the least possible permissions and redefine the app each time, then it definitely shouldn't be a part of this PR. I have a feeling what you might be thinking of (having two different sets of scopes for identity and integration) is not possible.

File: src/sentry/integrations/vsts/integration.py (hunk 14 above)
Comment (lauryndbrown) [reply src/sentry/integrations/vsts/integration.py]: No I haven't gotten any complaints.

File: src/sentry/identity/vsts/provider.py (hunk 13 above)
Comment (EvanPurkhiser) [reply src/sentry/identity/vsts/provider.py]: > The scopes must match the configuration setup in the application

Ah I see, in that case in the `identity_pipeline_config` you don't need to pass them again, it will just use the identity scopes if that config option is ommited

https://github.com/getsentry/sentry/blob/c62a7b251ee9fc840a81621374ba34bf46a7ceb5/src/sentry/identity/oauth2.py#L39-L40

Comment (lauryndbrown) [reply src/sentry/identity/vsts/provider.py]: Oh gotcha! Thanks! :)

File: src/sentry/integrations/vsts/integration.py [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+from sentry.pipeline import NestedPipelineView, PipelineView
+from sentry.identity.pipeline import IdentityProviderPipeline
+from sentry.utils.http import absolute_uri
+DESCRIPTION = """
+VSTS
+"""
+
+metadata = IntegrationMetadata(
+    description=DESCRIPTION.strip(),
+    author='The Sentry Team',
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/vsts/integration.py]: Recently added `noun` into this, see #8307

**Review by evanpurkhiser — APPROVED** [review]

**Comment by benvinegar** [comment]
Woot!
